Javascript console that exposes Filecoin API (#2492)
ChainSafe/forest: documentation/src/js_console.md

# Forest JavaScript Console

Forest console stands as an alternative to tools like [Curl](https://github.com/curl/curl) or other subcommands found in `forest-cli` for interacting with the Filecoin JSON-RPC API.

## Starting the console

`forest-cli attach` can be used to open a Javascript console connected to your Forest node.

Like for some other `forest-cli` subcommands you will need to pass an admin `--token` given what endpoints you will call.

For a description of different options please refer to the developer documentation [CLI page](https://github.com/ChainSafe/forest/blob/main/documentation/developer_documentation/CLI.md#cli).

## Interactive Use

First start Forest node inside another terminal:

```bash

> forest --chain calibnet --encrypt-keystore false

Reviewer: Let's not put this `--encrypt-keystore false` in the instructions. It's not required to run the JS console, and it's `true` for a reason.

```

To attach to your Forest node, run `forest-cli` with the `attach` subcommand:

```bash
forest-cli --token <TOKEN> attach 
```

You should now see a prompt and be able to interact:

```                                          
Welcome to the Forest Javascript console!

To exit, press ctrl-d or type :quit
> console.log("Forest running on " + chainGetName())
Forest running on calibnet
```

You can directly call JSON-RPC API endpoints that are bound to the console.

For example, `Filecoin.ChainGetName` is bound to the global `chainGetName` function.

### Tips

> - Console history is saved in your `~/.forest_history`
> - Use `:clear` to erase session commands before exiting

Reviewer: ```suggestion
- The console history is saved in your `~/.forest_history` after exiting.
- Use `:clear` to erase *current* session commands.
```

- Use `_BOA_VERSION` to get engine version

## Non-interactive Use

### Exec Mode

It is also possible to execute commands non-interactively by passing `--exec` flag and a JavaScript snippet to `forest-cli attach`. The result is displayed directly in the terminal rather than in the interactive console.

For example, to display the current epoch:

```bash
forest-cli attach --exec "syncStatus().ActiveSyncs[0].Epoch"
```

Or print node version:

```bash

> forest-cli attach --exec "commonVersion()"

Reviewer: ```
❯ target/release/forest-cli attach --exec "commonVersion()"
Uncaught ReferenceError: commonVersion is not defined
```

Author: At least there is no C++ like errors.

Reviewer: Yeeeeeeessss

```

## Builtins

### Helpers

Forest console comes with a number of helper functions that make interacting with Filecoin API easy:
 - `showPeers()`
 - `getPeer(peerID)`
 - `disconnectPeers(count)`
 - `isPeerConnected(peerID)`
 - `showWallet()`
 - `showSyncStatus()`
 - `sendFIL(to, attoAmount)`

### Timers

> In addition, to support part of the JavaScript language, the console also provides implementation for `sleep(seconds)` timer and a tipset based timer, `sleepTipsets(epochs)`, which sleeps till the number of new tipsets added is equal to or greater than `epochs`.

Reviewer: ```
> sleep(5)
Uncaught ReferenceError: sleep is not defined
> sleepTipsets(2)
Uncaught ReferenceError: sleepTipsets is not defined
```

Author: Sorry, this is only implemented in the shutdown PR (some kind of follow up). I can move the documentation there.

# Forest JavaScript Console

Forest console stands as an alternative to tools like [Curl](https://github.com/curl/curl) or other subcommands found in `forest-cli` for interacting with the Filecoin JSON-RPC API.

## Starting the console

`forest-cli attach` can be used to open a Javascript console connected to your Forest node.

Like for some other `forest-cli` subcommands you will need to pass an admin `--token` given what endpoints you will call.

For a description of different options please refer to the developer documentation CLI page.

## Interactive Use

First start Forest node inside another terminal:

```bash
./target/release/forest --chain calibnet --encrypt-keystore false
```

Copy the Admin token printed in the logs and set it to an environment variable, ie:

```bash

> TOKEN="..."

Author: I've updated the developer documentation with the `--token ` flag.

```

To attach to your Forest node, run `forest-cli` with the `attach` subcommand:

```bash
./target/release/forest-cli --token $TOKEN attach 
```

You should now see a prompt and be able to interact:

```                                          
Welcome to the Forest Javascript console!

To exit, press ctrl-d or type :quit
> console.log("Forest running on " + chainGetName())
Forest running on calibnet
```

> You can directly call JSON-RPC API endpoints that have been binded to the console.
> 
> For example `Filecoin.ChainGetName` will be binded to the global `chainGetName` function.

Reviewer: ```suggestion
You can directly call JSON-RPC API endpoints that are bound to the console.

For example, `Filecoin.ChainGetName` is bound to the global `chainGetName` function.
```

### Tips

- Console history is saved in your `~/.forest_history`

> - Use `:clear` to erase session commands before exiting

Reviewer: Where is it erased? It doesn't seem to erase the console history or clear the screen.

Author: It's erasing command of the current session.
Not global history (if you want to erase history from beginning you can wipe `~/.forest_history`.

Author: I agree, the wording could be better in the documentation.

- Use `_BOA_VERSION` to get engine version

## Non-interactive Use

### Exec Mode

It is also possible to execute commands non-interactively by passing `--exec` flag and a JavaScript snippet to `forest-cli attach`. The result is displayed directly in the terminal rather than in the interactive console.

For example, to display the current epoch:

```bash
./target/release/forest-cli attach --exec "syncStatus().ActiveSyncs[0].Epoch"
```

Or print node version:

```bash

> ./target/release/forest-cli attach --exec "commonVersion()"

Reviewer: this doesn't work for me, even with an admin token.

> ./target/release/forest-cli attach --exec "commonVersion()"

Reviewer: It'd be nice to be able to pass an entire JS file to `exec` or some other flag/subcommand

Author: Good idea but to keep this PR not to big, we can add address this in a future PR (maybe with some other enhancements, like command line completion etc).

In the meantime you can do:
```
$ ./forest-cli attach --exec "const _Foo = require('foo.js')"
hi
```
where foo.js:
```
console.log("hi");
```

```

## Builtins

### Helpers

Forest console comes with a number of helper functions that make interacting with Filecoin API easy:
 - `showPeers()`
 - `getPeer(peerID)`
 - `disconnectPeers(count)`
 - `isPeerConnected(peerID)`
 - `showWallet()`
 - `showSyncStatus()`
 - `sendFIL(to, attoAmount)`

### Timers

> In addition to support part of the JavaScript language, the console also provides implementation for `sleep(seconds)` timer and a tipset based timer, `sleepTipsets(epochs)` which sleeps till the number of new tipsets added is equal to or greater than `epochs`.
> 

Reviewer: ```suggestion
In addition, to support part of the JavaScript language, the console also provides implementation for `sleep(seconds)` timer and a tipset based timer, `sleepTipsets(epochs)`, which sleeps till the number of new tipsets added is equal to or greater than `epochs`.

```

### Modules

CommonJS modules is the way to package JavaScript code for Forest console. You can import modules using the `require` function:

```bash
./target/release/forest-cli attach --exec "const Math = require('calc'); console.log(Math.add(39,3))"
```

where `calc.js` is:

```javascript
module.exports = {
    add: function (a, b) {
      return a + b;
    },
    multiply: function (a, b) {
      return a * b;
    },
};
```

By default modules will be loaded from the current directory. Use `--jspath` flag to indicate another path.

## Limitations

Forest's console is built using [Boa Javascript engine](https://github.com/boa-dev/boa). It does support promises or `async` functions, but keep in mind that Boa is not fully compatible with ECMAScript yet.

> Not every endpoints from the Filecoin API has been binded to the console at the moment.

Reviewer: ```suggestion
Not every endpoint from the Filecoin API has been bound to the console. Please [create an issue](https://github.com/ChainSafe/forest/issues) if you need one that is not available.
```
